### PR TITLE
planner: fix index merge intersection case will have redundant partial path

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/50496

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: pkg/planner/core/indexmerge_path.go [hunk 1]
```diff
@@ -794,9 +816,35 @@ func (ds *DataSource) generateMVIndexMergePartialPaths4And(normalPathCnt int, in
...
+			if rec.countAfterAccess > countAfterAccess {
+				mm[allHashCodesKey] = &record{idx, partialPaths4ThisMvIndex, countAfterAccess}
+			}
+		}
+	}
+	// after all mv index is traversed, pick those remained paths which has already been de-duplicated for its accessFilters.
+	recordsCollection := make([]*record, 0, len(mm))
+	for _, one := range mm {
+		recordsCollection = append(recordsCollection, one)
+	}
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_path.go:841]:
```suggestion
	recordsCollection := maps.Values(mm)
```
And import `"golang.org/x/exp/maps"`

File: pkg/planner/core/indexmerge_path.go [hunk 2]
```diff
@@ -794,9 +816,35 @@ func (ds *DataSource) generateMVIndexMergePartialPaths4And(normalPathCnt int, in
...
+		// sort allHashCodes to make the unified hashcode for slices of all accessFilters.
+		slices.Sort(allHashCodes)
+		allHashCodesKey := strings.Join(allHashCodes, "")
+		if rec, ok := mm[allHashCodesKey]; !ok {
+			// compute the count after access from those intersection partial paths, for this mv index usage.
+			countAfterAccess := float64(histColl.RealtimeCount) * cardinality.CalcTotalSelectivityForMVIdxPath(histColl, partialPaths4ThisMvIndex, true)
+			mm[allHashCodesKey] = &record{idx, partialPaths4ThisMvIndex, countAfterAccess}
+		} else {
+			// pick the minimum countAfterAccess's paths.
+			countAfterAccess := float64(histColl.RealtimeCount) * cardinality.CalcTotalSelectivityForMVIdxPath(histColl, partialPaths4ThisMvIndex, true)
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_path.go:831]: Unnecessary duplication with L827.

File: tests/integrationtest/t/planner/core/casetest/index/index.test [hunk 3]
```diff
@@ -285,3 +285,8 @@ CREATE TABLE `t2` (`a` float NOT NULL DEFAULT '5217.6055',`b` json NOT NULL,`c`
 EXPLAIN SELECT /*+ inl_join(t1)*/ `t2`.`c` AS `r0` FROM `t1` JOIN `t2` ON `t1`.`a`=`t2`.`d` WHERE `t2`.`d`='' AND NOT (6252179388429456273 MEMBER OF (`t1`.`c`));
 SHOW WARNINGS
 
+# TestIndexMergeEliminateRedundantAndPaths
+DROP table if exists items;
+CREATE TABLE `items` (`item_primary_key` varbinary(255) NOT NULL,`normalized_short_link_clusters` json DEFAULT NULL,`feed_profile_ids` json DEFAULT NULL,`point_of_sale_country` varchar(2) DEFAULT NULL,PRIMARY KEY (`item_primary_key`) /*T![clustered_index] CLUSTERED */,KEY `feed_profile_ids` ((cast(`feed_profile_ids` as unsigned array))),KEY `normalized_short_link_clusters` ((cast(`normalized_short_link_clusters` as char(1000) array)),`point_of_sale_country`),KEY `normalized_short_link_clusters_old` ((cast(`normalized_short_link_clusters` as char(1000) array))));
+EXPLAIN SELECT /*+ use_index_merge(items, feed_profile_ids, normalized_short_link_clusters_old, normalized_short_link_clusters) */ * FROM   items WHERE   15975127 member of (feed_profile_ids)   AND "OC8p0106XTkt.org/s/link" member of (normalized_short_link_clusters) LIMIT   10;
```
Comment (time-and-fate) [inline tests/integrationtest/t/planner/core/casetest/index/index.test:291]:
```suggestion
EXPLAIN FORMAT = brief SELECT /*+ use_index_merge(items, feed_profile_ids, normalized_short_link_clusters_old, normalized_short_link_clusters) */ * FROM   items WHERE   15975127 member of (feed_profile_ids)   AND "OC8p0106XTkt.org/s/link" member of (normalized_short_link_clusters) LIMIT   10;
```

File: tests/integrationtest/t/planner/core/casetest/index/index.test [hunk 4]
```diff
@@ -285,3 +285,8 @@ CREATE TABLE `t2` (`a` float NOT NULL DEFAULT '5217.6055',`b` json NOT NULL,`c`
 EXPLAIN SELECT /*+ inl_join(t1)*/ `t2`.`c` AS `r0` FROM `t1` JOIN `t2` ON `t1`.`a`=`t2`.`d` WHERE `t2`.`d`='' AND NOT (6252179388429456273 MEMBER OF (`t1`.`c`));
 SHOW WARNINGS
 
+# TestIndexMergeEliminateRedundantAndPaths
+DROP table if exists items;
+CREATE TABLE `items` (`item_primary_key` varbinary(255) NOT NULL,`normalized_short_link_clusters` json DEFAULT NULL,`feed_profile_ids` json DEFAULT NULL,`point_of_sale_country` varchar(2) DEFAULT NULL,PRIMARY KEY (`item_primary_key`) /*T![clustered_index] CLUSTERED */,KEY `feed_profile_ids` ((cast(`feed_profile_ids` as unsigned array))),KEY `normalized_short_link_clusters` ((cast(`normalized_short_link_clusters` as char(1000) array)),`point_of_sale_country`),KEY `normalized_short_link_clusters_old` ((cast(`normalized_short_link_clusters` as char(1000) array))));
```
Comment (time-and-fate) [inline tests/integrationtest/t/planner/core/casetest/index/index.test:290]: I suggest we change the table and column names.

File: pkg/planner/core/indexmerge_path.go [hunk 5]
```diff
@@ -761,6 +762,14 @@ func (ds *DataSource) generateMVIndexMergePartialPaths4And(normalPathCnt int, in
...
 	mvAndPartialPath := make([]*util.AccessPath, 0, len(possibleMVIndexPaths))
 	usedAccessCondsMap := make(map[string]expression.Expression, len(indexMergeConds))
+	// fill the possible indexMergeConds down to possible index merge paths. If two index merge path
+	// share same accessFilters, pick the one with minimum countAfterAccess.
+	type record struct {
+		originOffset     int
+		paths            []*util.AccessPath
+		countAfterAccess float64
+	}
+	mm := make(map[string]*record, 0)
```
Comment (time-and-fate) [inline pkg/planner/core/indexmerge_path.go:772]: Let's use a more meaningful variable name or add some comments about what this map is for.

Comment (time-and-fate) [reply pkg/planner/core/indexmerge_path.go:772]: And the `, 0` is actually redundant.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**